hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -321,65 +293,64 @@ void completeJob(long jobId, String coordinator, long completionTime, Throwable
-        validJobIds.addAll(runningJobIds);
-        validJobIds.addAll(jobRecords.keySet());
-
-        // Job ids are never cleaned up.
-        // We also don't clean up job records here because they might be started in parallel
-        // while cleanup is running. If a job id is not running or is completed, it might be
-        // suitable to clean up job resources.
-        randomIds.keySet(new FilterJobIdPredicate())
-                 .stream()
-                 .filter(jobId -> !validJobIds.contains(jobId))
-                 .forEach(jobId -> {
-                     IMap<String, Object> resources = getJobResources(jobId);
-                     EntryView<String, Object> marker = resources.getEntryView(RESOURCE_MARKER);
-                     // If the marker is absent, then job resources may be still uploaded.
-                     // Just put the marker so that the job resources may be cleaned up eventually.
-                     // If the job resources are still being uploaded, then the marker will be overwritten, which is ok.
-                     if (marker == null) {
-                         resources.putIfAbsent(RESOURCE_MARKER, System.currentTimeMillis());
-                     } else if (isMarkerExpired(marker)) {
-                         // The marker has been around for defined expiry time and the job still wasn't started.
-                         // We assume the job submission was interrupted - let's clean up the data.
-                         cleanupJobResourcesAndSnapshots(jobId, resources);
-                     }
-                 });
-    }
-
-    private boolean isMarkerExpired(EntryView<String, Object> record) {
-        return (System.currentTimeMillis() - (Long) record.getValue()) >= resourcesExpirationMillis;
-    }
-
-    List<JobRecord> getJobRecords(String name) {
-        return jobRecords.values(new FilterJobRecordByNamePredicate(name)).stream()
-                         .sorted(comparing(JobRecord::getCreationTime).reversed()).collect(toList());
+    void cleanup(NodeEngine nodeEngine) {
+        Collection<DistributedObject> maps =
+                nodeEngine.getProxyService().getDistributedObjects(MapService.SERVICE_NAME);
+
+        maps.stream()
+                .filter(map -> map.getName().startsWith(SNAPSHOT_DATA_MAP_PREFIX))
+                .forEach(map -> {
+                    long id = jobIdFromMapName(map.getName(), SNAPSHOT_DATA_MAP_PREFIX);
+                    // job is not active
+                    if (!jobRecords.containsKey(id)) {
+                        map.destroy();
+                    }
+                });
+
+        maps.stream()
+                .filter(map -> map.getName().startsWith(RESOURCES_MAP_NAME_PREFIX))
+                .forEach(map -> {
+                    long id = jobIdFromMapName(map.getName(), RESOURCES_MAP_NAME_PREFIX);
+                    // if job is finished, we can safely delete the map
+                    if (jobResults.containsKey(id)) {
```

Comment:
Again, pre-loading keys might be better. Except if there are few active jobs and lots of completed jobs. Maybe:
```java
Set<Long> jobResultsKeysLocal = jobResults.size() > 50 * jobRecordKeysLocal.size()
    ? Collections.newSetFromMap(jobResults) : jobResults.keySet()
```

Reply:
I don't think it's necessary, in the "happy path" the job is likely to be already running and the map won't be queried.